src/Microsoft.ML.AutoML/Experiment/Experiment.cs

```diff
@@ -49,23 +53,55 @@ internal class Experiment<TRunDetail, TMetrics> where TRunDetail : RunDetail
             _datasetColumnInfo = datasetColumnInfo;
             _runner = runner;
             _logger = logger;
+            _experimentTimerExpired = false;
+        }
+
+        private void MaxExperimentTimeExpiredEvent(object sender, EventArgs e)
+        {
+            // If at least one model was run, end experiment immediately.
+            // Else, wait for first model to run before experiment is concluded.
+            _experimentTimerExpired = true;
+            if (_history.Any(r => r.RunSucceeded))
+            {
+                _logger.Warning("Allocated time for Experiment of {0} seconds has elapsed with {1} models run. Ending experiment...",
+                    _experimentSettings.MaxExperimentTimeInSeconds, _history.Count());
+                _currentModelMLContext.CancelExecution();
+            }
         }
 
         public IList<TRunDetail> Execute()
         {
-            var stopwatch = Stopwatch.StartNew();
             var iterationResults = new List<TRunDetail>();
+            // Create a timer for the max duration of experiment. When given time has
+            // elapsed, MaxExperimentTimeExpiredEvent is called to interrupt training
+            // of current model. Timer is not used if no experiment time is given, or
+            // is not a positive number.
+            if (_experimentSettings.MaxExperimentTimeInSeconds > 0)
+            {
+                Timer timer = new Timer(_experimentSettings.MaxExperimentTimeInSeconds * 1000);
+                timer.Elapsed += MaxExperimentTimeExpiredEvent;
+                timer.AutoReset = false;
+                timer.Enabled = true;
+            }
+            // If given max duration of experiment is 0, only 1 model will be trained.
+            // _experimentSettings.MaxExperimentTimeInSeconds is of type uint, it is
+            // either 0 or >0.
+            else
+                _experimentTimerExpired = true;
 
             do
             {
                 var iterationStopwatch = Stopwatch.StartNew();
 
                 // get next pipeline
                 var getPipelineStopwatch = Stopwatch.StartNew();
-                var pipeline = PipelineSuggester.GetNextInferredPipeline(_context, _history, _datasetColumnInfo, _task,
-                    _optimizingMetricInfo.IsMaximizing, _experimentSettings.CacheBeforeTrainer, _trainerAllowList);
 
-                var pipelineInferenceTimeInSeconds = getPipelineStopwatch.Elapsed.TotalSeconds;
+                // A new MLContext is needed per model run. When max experiment time is reached, each used
+                // context is canceled to stop further model training. The cancellation of the main MLContext
+                // a user has instantiated is not desirable, thus additional MLContexts are used.
+                _currentModelMLContext = new MLContext(((ISeededEnvironment)_context.Model.GetEnvironment()).Seed);
```

Comment:
Won't this use the same Seed at the beginning of each iteration?

Reply:
What good is the `_context`? What is not desired about cancelling the main MLContext?

Reply:
Thanks for hopping on the code review. 

The main `_context` is for when the _user_ canceled their ML workflow. The purpose of the `_currentModelMLContext` is only to stop the current model being trained, when time expires, and return back to the user code.

If we cancel the main MLContext, the user's subsequent steps like refitting the estimator will not work after you return from the AutoML call.

Normal AutoML use looks like:
1. Load data
2. Call AutoML
3. Score test dataset
4. Refit estimator on all available data (thrice training)
5. ...any other steps in the user's ML workflow... perhaps model stacking?

Just because the time expired for the AutoML job in step (2), doesn't mean the user doesn't need the next steps.

Reply:
> Won't this use the same Seed at the beginning of each iteration?

Yes, and it is slightly different behavior than current. Previously it could just continue pulling new random numbers from the PRNG; now it will reset between iterations to the same steam of random numbers. The main difference I see is this causes deterministic splitting of the dataset for all iterations in the sweep (if the user sets a seed). This lowers the variability within the sweep iterations, which is a bit less optimal for the Bayesian hyperparamameter optimization, as it can otherwise overfit the repeating splits it sees.

Perhaps we should make a new PRNG seeded on the `_context` seed, and feed the next random as the seed to each new MLContext? This gives the best of both worlds, (1) the user has a now more deterministic run if the call AutoML with the same seed, as the splits are now regenerated identically for each AutoML re-run, and (2) the variability between iterations is maintained.